Support report of notify area and fd

shahafsh opened the pull request.

shahafsh committed:
net/mlx5_vdpa: report virtio notify area

Select the notify area for the given virtqueue doorbells. The offset
selected is relative to the base address of the virtio doorbells on the
device PCI bar.

The current KVM and QEMU code only supports mapping of areas in size of
page, as it much fits into the way KVM register the EPT/Shadow page
table entries.

This is OK for BlueField1 solution, however, as an action item will need to
check if we can modify KVM and QEMU and support smaller sizes.

Signed-off-by: Shahaf Shuler <shahafs@mellanox.com>

shahafsh committed:
net/mlx5_vdpa: report vDPA device fd

This is the verbs command fd.

Signed-off-by: Shahaf Shuler <shahafs@mellanox.com>

shahafsh committed:
net/mlx5_vdpa: report extra features

Enable extra vDPA and protocol featrures in order to be able to
report the host notifier.

Signed-off-by: Shahaf Shuler <shahafs@mellanox.com>

shamoya commented on drivers/net/mlx5/mlx5_vdpa.c:
```diff
@@ -18,22 +18,30 @@
+			    (1ULL << VIRTIO_F_VERSION_1) | \
+			    (1ULL << VHOST_USER_PROTOCOL_F_SLAVE_REQ) | \
+			    (1ULL << VHOST_USER_PROTOCOL_F_SLAVE_SEND_FD) | \
+			    (1ULL << VHOST_USER_PROTOCOL_F_HOST_NOTIFIER));
```
Those last 3 (VHOST_USER_PROTOCOL*) are VHOST protocol features and not VIRTIO features, so they need to go in a separate 64 bitmask, as we need to return them from get_protocol_features (and not from get_features).
The first one VHOST_USER_F_PROTOCOL_FEATURES is indeed in the VIRTIO features to negotiate if protocol features are supported at all.

shahafsh replied:
OK. will adjust.

shamoya commented on drivers/net/mlx5/mlx5_vdpa.c:
```diff
@@ -18,22 +18,30 @@
 	uint16_t max_num_virtqs;
 	uint64_t virtio_net_features;
+	uint16_t notify_base; /* notify area base offset. */
+	uint16_t notify_size; /* notify area size. */
```
According to our PRM sketch, there is also notify_stride

shahafsh replied:
OK.

shamoya commented on drivers/net/mlx5/mlx5_vdpa.c:
```diff
@@ -147,14 +141,81 @@ mlx5_vdpa_query_virtio_caps(struct vdpa_priv *priv)
+	return (((offset_cap >> MLX5_IB_MMAP_CMD_SHIFT) << ext_offset) |
+		(MLX5_IB_MMAP_VIRTIO_NOTIFY << MLX5_IB_MMAP_CMD_SHIFT) |
+		(offset_cap & MLX5_IB_MMAP_CMD_MASK));
+}
```
We need to take the queue index here to pick the right stride inside the bar.
Also I don't understand how this can work. the cmd_fd is a UAR page, and offset cap you get from FW
is absolute offset from start of BAR0.
So we need to assume we have a new fd which is the correct page from the BAR, and then you need to set the offset from within the page according to the queue index. 
Right?

shahafsh replied:
This is not correct. the cmd_fd maps the device bar starting from the initialization segment.

shamoya commented:
@shahafsh - there is a kernel patch with it right? 
If so, add it here or its path

shahafsh commented:
@shamoya see updated pr.

sent you also the linux and qemu patches by mail.
let's have a call if you still have concerns.

shamoya commented on drivers/net/mlx5/mlx5_vdpa.c:
```diff
@@ -147,14 +160,92 @@ mlx5_vdpa_query_virtio_caps(struct vdpa_priv *priv)
 }
 
+#define MLX5_IB_MMAP_CMD_SHIFT 8
+#define MLX5_IB_MMAP_CMD_MASK ((1 << MLX5_IB_MMAP_CMD_SHIFT) - 1)
```
Hi @shahafsh 
should be ((1 << MLX5_IB_CMD_SIZE) - 1) right ? 
It happens to be they are both 8 bits but it's confusing.

shahafsh replied:
No.

the layout of the mmap offset is:
page_offset[15:8] | mmap_command[7:0] | page_offset[7:0]

or w/ macro write
page_offset[15:MLX5_IB_MMAP_CMD_SHIFT] | mmap_command[(MLX5_IB_CMD_SIZE-1):0] | page_offset[(MLX5_IB_MMAP_CMD_SHIFT-1):0]

i think the name MLX5_IB_MMAP_CMD_MASK is confusing. will change it.

shamoya commented on drivers/net/mlx5/mlx5_vdpa.c:
```diff
@@ -147,14 +160,92 @@ mlx5_vdpa_query_virtio_caps(struct vdpa_priv *priv)
+#define MLX5_IB_MMAP_CMD_SHIFT 8
+#define MLX5_IB_MMAP_CMD_MASK ((1 << MLX5_IB_MMAP_CMD_SHIFT) - 1)
+#define MLX5_IB_CMD_SIZE 8
+#define MLX5_IB_MMAP_VIRTIO_NOTIFY 9
```
This will eventually be replaced with MLX5_IB_MMAP_VIRTIO_NOTIFY right ?
If so, maybe we can add TODO here

shahafsh replied:
hmm, if indeed the kernel will decided to go w/ this approach we will need to add ifdef, like
#ifndef MLX5_IB_MMAP_VIRTIO_NOTIFY 
#define MLX5_IB_MMAP_VIRTIO_NOTIFY 9
#endif

because the linux header not all distro/customer updates the linux user headers when upgrading the kernel.

at this point it is just a temporary code. if indeed the same code will go to upstream then we will adjust.

shamoya commented on drivers/net/mlx5/mlx5_vdpa.c:
```diff
@@ -147,14 +160,92 @@ mlx5_vdpa_query_virtio_caps(struct vdpa_priv *priv)
+	if (!list)
+		goto error;
+	notify_stride = list->priv->caps.notify_stride;
+	*offset = 0 | mlx5_vdpa_get_notify_offset(qid * notify_stride);
```
why do we need the "0 |"?
It makes the code less readable IMO

shahafsh replied:
just to not rely on the compiler to clear the rest of the offset bits. 

i can remove it.

shamoya commented on drivers/net/mlx5/mlx5_vdpa.c:
```diff
@@ -147,14 +160,92 @@ mlx5_vdpa_query_virtio_caps(struct vdpa_priv *priv)
+	uint16_t ext_offset = MLX5_IB_MMAP_CMD_SHIFT + MLX5_IB_CMD_SIZE;
+	return (((offset >> MLX5_IB_MMAP_CMD_SHIFT) << ext_offset) |
+		(MLX5_IB_MMAP_VIRTIO_NOTIFY << MLX5_IB_MMAP_CMD_SHIFT) |
+		(offset & MLX5_IB_MMAP_CMD_MASK));
```
This function is not readable enough, very hard to read this piece of code and understand what it does.
Please consider making it more simple and layered, like in libmlx5, for instance.

offset = 0;
set_command(MLX5_IB_MMAP_VIRTIO_NOTIFY, &offset);
set_extended_index(offset, &offset)
.
.
.

shahafsh replied:
yes, i didn't think that it justify to have such comprehensive APIs only for setting a single command. 

i can change it, no problem.

shamoya commented on drivers/net/mlx5/mlx5_vdpa.c:
```diff
@@ -147,14 +160,92 @@ mlx5_vdpa_query_virtio_caps(struct vdpa_priv *priv)
+	 * TODO (shahafs): Take from QUERY HCA CAP Device Emulation Capabilities.
+	 * For now only set protocol features support
+	 */
+	priv->caps.notify_stride = MLX5_VDPA_SW_NOTIFY_STRIDE;
```
Once you'll rebase, you'll see there's already piece of code that try to query the VIRTIO params from FW.
Plz Just add there as well to get the notify_stride (and in the else "FW no support" have this).
Also plz note the stride should come in a log from FW.

shahafsh replied:
Thanks. 

on the next version i removed the stride completely. the support for doorbell offset not page aligned is not yet ready for KVM. am planning to look into it. 

if we can make it work, will add the support.

Note that we consider to change ConnectX-6Dx doorbell layout due to it as well.

shamoya reviewed (commented):
Nice @shahafsh 
See my small comments

shahafsh committed:
mlx5_vdpa: support notify relay

So that we can support virtio v0.95.
A note for upstream submission - move this code into a shared location
on vhost lib.

Signed-off-by: Shahaf Shuler <shahafs@mellanox.com>

shahafsh commented:
@shamoya 

thanks for comments. 
see revised version w/ your comments addressed.

shamoya reviewed (approved):
Looks great @shahafsh 
Thanks !
